grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,108 @@
+
+        // If the invoice item is for an adjustment.
+        OrderAdjustment orderAdjustment = prepaymentInvoiceItem.orderAdjustment
+        if (orderAdjustment) {
+            return (orderAdjustment?.canceled ? 0 : orderAdjustment.totalAdjustments) * prepaymentPercent
+        }
+
+        // Else if the invoice item is for a canceled order item.
+        OrderItem orderItem = prepaymentInvoiceItem.orderItem
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            return 0
+        }
+
+        // Else the invoice item is for a non-canceled order item.
+        return (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0) * prepaymentPercent
+    }
+
+    /**
+     * For STEP 2 of the migration.
+     *
+     * For every item in a given prepayment invoice, computes the inverse invoice item and insert it to the database
+     * by adding it to the given final invoice.
+     *
+     * This method assumes that because there is both a prepayment invoice and a final invoice, that the order
+     * associated with both invoices has been fully invoiced, and so there's no need to check each item of the
+     * prepayment invoice individually, they should ALL need to have inverse items created for them.
+     */
+    void generateInverseInvoiceItems(Invoice prepaymentInvoice, Invoice finalInvoice) {
+        prepaymentInvoice.invoiceItems.each { InvoiceItem prepaymentInvoiceItem ->
+            List<InvoiceItem> inverseItems = createInverseInvoiceItems(prepaymentInvoiceItem, finalInvoice)
+            inverseItems?.each { finalInvoice.addToInvoiceItems(it) }
+        }
+        finalInvoice.save(failOnError: true, flush: true)
+    }
+
+    private List<InvoiceItem> createInverseInvoiceItems(InvoiceItem prepaymentInvoiceItem, Invoice finalInvoice) {
+        // If the invoice item is for an adjustment.
+        OrderAdjustment orderAdjustment = prepaymentInvoiceItem.orderAdjustment
+        if (orderAdjustment) {
+            InvoiceItem inverseItem = prepaymentInvoiceService.createInverseItemForOrderAdjustment(orderAdjustment)
+            return inverseItem ? [inverseItem] : []
+        }
+
+        // Else if the invoice item is for a canceled order item.
+        OrderItem orderItem = prepaymentInvoiceItem.orderItem
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            InvoiceItem inverseItem = prepaymentInvoiceService.createInverseItemForCanceledOrderItem(orderItem)
+            return inverseItem ? [inverseItem] : []
+        }
+
+        // Else the invoice item is for a non-canceled order item. The item might be split across multiple shipments
+        // so we need to loop the shipment items and generate an inverse for each.
+        // TODO: This was copied from PrepaymentInvoiceService. Why do we need to do this? Shouldn't it be a single inverse for each prepay, regardless of if it was split over multiple shipments?
```

Comment:
@awalkowiak highlighting this for discussion

Reply:
Partial invoicing had to made from the shipment items, but for this case, where we have only one final invoice, I think we could just make 100% copy prepayment item (well actually, 99%, because amount will multiplied by `-1`, and quantity 0 for canceled items)

Reply:
good point. Should quantity be 0 though? Looking at `PrepaymentInvoiceService.createInverseItemForCanceledOrderItem` I see it does `inverseItem.quantity = prepaymentItem.quantity`

https://github.com/openboxes/openboxes/blob/feature/OBPIH-6398-partial-invoices-for-prepaid-po/grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy#L164